lib/Sema/CSDiagnostics.h

```diff
@@ -42,19 +43,17 @@ class FunctionArgApplyInfo;
 class FailureDiagnostic {
   const Solution &S;
   ConstraintLocator *Locator;
-  DiagnosticBehavior behaviorLimit;
+  FixBehavior fixBehavior;
 
 public:
   FailureDiagnostic(const Solution &solution, ConstraintLocator *locator,
-                    DiagnosticBehavior behaviorLimit =
-                        DiagnosticBehavior::Unspecified)
-      : S(solution), Locator(locator), behaviorLimit(behaviorLimit) {}
+                    FixBehavior fixBehavior = FixBehavior::Error)
```

Comment:
I think we can avoid mixing fixes and diagnostics here. How about we bring back `behaviorLimit` but make it `Optional<DiagnosticBehavior>` which is going to be `None` by default, and it's going to be supplied by the fix that constructs a diagnostic via call to `limitsDiagnosticBehaviorTo()`?

Reply:
Because we want `isFatal` and `impact` to operate on `FixBehavior`, we need to carry `FixBehavior` through here and map it down to `DiagnosticBehavior` only at the point of diagnostic emission.

Reply:
I'm not exactly sure what you mean - `FailureDiagnostic` represents a diagnostic emission for a `ConstraintFix` point of view, so the fix should tell it whether it wants to limit the behavior or not, it shouldn't be up to diagnostic to decide what to do based on the fix behavior that's why I suggested to package that logic into `ConstraintFix::limitsDiagnosticBehaviorTo()` which is going to produce `Optional< DiagnosticBehavior>`.